src/notebook.c

```diff
@@ -65,6 +66,8 @@ static gboolean switch_in_progress = FALSE;
 static GtkWidget *switch_dialog = NULL;
 static GtkWidget *switch_dialog_label = NULL;
 
+static gboolean doc_saves_to_new_file = FALSE;
```

Comment:
> Do you mean to have something like `g_signal_connect(geany_object, "document-open", G_CALLBACK(on_document_open), &doc_saves_to_new_file);`? 

yes

> Should we really do that?

I don't know.  Ideally there wouldn't be any need for a hack to know whether it was saved or saved as…
BTW your technique here doesn't really work: it only finds out whether it's the first save, not whether the filename changed.